### PR TITLE
Add Yarn-related rules

### LINKED ISSUE #11: Add checks for Yarn compliance
Since the module template uses Yarn v3, we want all projects to conform to the same thing. So, minimally, for this ticket, we want this tool to be able to look a project and verify whether it is using the same version of Yarn as is specified in the module template.

There are several ways to check that the Yarn version conforms, though. If this is the case for a project, then these things will be true:

- The `packageManager` field in `package.json` should exist
- The value of the `packageManager` field should match the same value as the module template
- `.yarn/plugins/@yarnpkg/plugin-allow-scripts.cjs` should exist, and its contents should match what's in the module template
- `.yarn/plugins/@yarnpkg/plugin-constraints.cjs` should exist, and its contents should match what's in the module template
- `.yarn/releases/yarn-3.2.1.cjs` should exist, and its contents should match what's in the module template
- The `.yarnrc.yml` file should exist, and the contents of this file should match the same file in the module template
- The README should include a line `- Install [Yarn v3](https://yarnpkg.com/getting-started/install)`

We can create 7 rules to encapsulate all of these checks.

### LINKED ISSUE #23: module-lint: Make sure package.json exists
As we are planning to add multiple checks (for example yarn compliance), before that, we need to make sure package.json exists in the project.

### PR DESCRIPTION
**Pull request by mcmire**
For a given project, we want to ensure that it is using the same Yarn version as the module template, that Yarn is configured exactly the same way as in the module template, and that all of the Yarn-related files are present, including the Yarn "binary" in `.yarn/releases` and the plugins in `.yarn/plugins`. We also want to ensure that the version we recommend users install in the README matches the same version as the module template.

To accomplish this, this commit adds six new rules:

- `all-yarn-modern-files-conform`
  - This checks that `.yarnrc.yml` is present and matches the module template, and any files in the template's `.yarn/plugins` and `.yarn/releases` directories are present in the project and match.
- `classic-yarn-config-file-absent`
  - This checks that `.yarnrc`, which is the configuration file for Yarn Classic, is not present.
- `package-manager-conforms`
  - This checks that the `packageManager` field in the project's `package.json` is the same as that of the module template's.
- `readme-lists-correct-yarn-version`
  - This checks that the Yarn version in the project's README is the same as that of the module template's.
- `require-readme`
  - This checks that the project has a README.
- `require-valid-package-manifest`
  - This checks the project has a `package.json` and that the JSON-parsed content of this file matches a known shape.

Fixes #11.
Fixes #23.

### CODE REVIEW

File: src/misc-utils.ts [hunk 1]
```diff
@@ -26,6 +46,148 @@ export async function getEntryStats(
...
+  } catch (error) {
+    throw wrapError(error, `Could not read directory '${directoryPath}'`);
+  }
+}
+
+/**
+ * Builds a new error object, linking it to the original error via the `cause`
+ * property if it is an Error.
+ *
+ * This is different from the `@metamask/utils` version because it allows for
```
Comment (Gudahtt) [inline src/misc-utils.ts:130]: Nit: Have you considered using the `wrapError` function from utils, and overwriting the error code on the resulting error? Most of the logic is still the same, that would let us get rid of most of this complexity.

File: src/misc-utils.ts [hunk 2]
```diff
@@ -1,6 +1,26 @@
...
+import type { Json } from '@metamask/utils/node';
+import fs from 'fs';
+import path from 'path';
+import { ErrorWithCause } from 'pony-cause';
+import { StructError, assert } from 'superstruct';
+import type { Struct } from 'superstruct';
+import type { ObjectSchema } from 'superstruct/dist/utils';
+
+/**
+ * Represents a directory entry. Like fs.Dirent, but includes a `path` property
```
Comment (Gudahtt) [inline src/misc-utils.ts:11]: Nit: It seems like the `path` property is unchanged from the `Dirent` object, but rather there are _two additional_ properties for the "full" (maybe should be called "absolute"?) and relative path

File: src/misc-utils.ts (hunk 1 above)
Comment (Gudahtt) [reply src/misc-utils.ts:130]: Nice feature though, seems useful to add to the `utils` version of this function.

File: src/main.test.ts [hunk 3]
```diff
@@ -215,54 +281,93 @@ Elapsed time:  0 ms
...
+Results:       8 passed, 0 failed, 8 total
 Elapsed time:  0 ms
 
 `,
         );
       });
     });
 
-    it('lists the rules executed against the default repositories which fail', async () => {
+    it('lists the failing rules executed against the default repositories', async () => {
```
Comment (Gudahtt) [inline src/main.test.ts:370]: Question, what is this test intended to do? Is it just demonstrating that it's possible for rules to fail? I'm wondering how it's different from the "passing" test; the only difference seems to be the number of failures.

File: src/rules/helpers.test.ts [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+} from '@metamask/utils/node';
+import path from 'path';
+
+import { directoryExists, fileExists } from './helpers';
+import { buildMetaMaskRepository, withinSandbox } from '../../tests/helpers';
+import { fail, pass } from '../execute-rules';
+
+describe('fileExists', () => {
+  it('passes if the given path refers to an existing file', async () => {
+    expect.assertions(1);
```
Comment (Gudahtt) [inline src/rules/helpers.test.ts:13]: What motivated this? We don't usually bother stating this explicitly.

File: src/rules/helpers.ts [hunk 5]
```diff
@@ -53,3 +54,165 @@ export function buildRule<Name extends RuleName>({
...
+/**
+ * A rule utility which determines not only whether a file that's assumed to
+ * exist in the template exists in the project as well, but also whether it
+ * matches the same file in the template content-wise.
+ *
+ * @param filePath - The path to a file in both the template and project.
+ * @param ruleExecutionArguments - Rule execution arguments.
+ * @returns Either a successful or failed result.
+ */
+export async function fileConforms(
```
Comment (Gudahtt) [inline src/rules/helpers.ts:133]: These last three functions appear to be missing tests

File: src/rules/readme-lists-correct-yarn-version.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    'Does the README conform by recommending the correct Yarn version to install?',
+  dependencies: [RuleName.RequireReadme, RuleName.AllYarnModernFilesConform],
+  execute: async ({ template, project, pass, fail }) => {
+    const entryPath = 'README.md';
+
+    const fileContentInTemplate = await template.fs.readFile(entryPath);
+    const fileContentInProject = await project.fs.readFile(entryPath);
+
+    const match = fileContentInTemplate.match(
+      /Install \[Yarn [^[\]]+\]\([^()]+\)/u,
```
Comment (Gudahtt) [inline src/rules/readme-lists-correct-yarn-version.ts:16]: Nit: Might be simpler to write it this way, using a non-greedy modifier:

```suggestion
      /Install \[Yarn .+?\]\(.+?\)/u,
```

At least I think that should do the same thing

**Review by Gudahtt — COMMENTED** [review]
Looks great! Left a few comments, mostly minor suggestions or questions though

File: src/main.test.ts (hunk 3 above)
Comment (mcmire) [reply src/main.test.ts:370]: Hmm, good question. My thought here was that the "passing" test would exercise a maximal scenario, and the "failing" test would exercise a minimal scenario. That is, in the first case, all of the rules are satisfied, so all of them should be passing, and in the second case, none of them are satisfied, so none of them should be passing, and error messages should be shown for all of them. In both cases, we also test that if a rule has children and the rule fails, then none of the children are run.

Comment (Gudahtt) [reply src/main.test.ts:370]: Oh I see. In that case, there is one test that is passing even in this "failing" scenario (the one about not having the old yarn v1 config around)

File: src/misc-utils.ts (hunk 1 above)
Comment (mcmire) [reply src/misc-utils.ts:136]: > Nit: Have you considered using the `wrapError` function from utils, and overwriting the error code on the resulting error? Most of the logic is still the same, that would let us get rid of most of this complexity.

Good call — I've simplified it here: [52e9ce7](https://github.com/MetaMask/module-lint/pull/39/commits/52e9ce70d073c8f95cfe8933af068afcc3e89ccf)

File: src/misc-utils.ts (hunk 2 above)
Comment (mcmire) [reply src/misc-utils.ts]: Ah, true, thanks. JSDoc updated here: [685621e](https://github.com/MetaMask/module-lint/pull/39/commits/685621e04ee299e7055b2d5f24288a22c71036d9)

File: src/rules/all-yarn-modern-files-conform.ts [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+import { RuleName } from './types';
+
+export default buildRule({
+  name: RuleName.AllYarnModernFilesConform,
+  description:
+    'Are all of the files for Yarn Modern present, and do they conform?',
+  dependencies: [],
+  execute: async (ruleExecutionArguments) => {
+    const results = await Promise.all([
+      fileConforms('.yarnrc.yml', ruleExecutionArguments),
```
[the remark this answers is not in the document]
Comment (mcmire) [reply src/rules/all-yarn-modern-files-conform.ts:16]: Hmm, interesting. So, I'm not sure. I do think that anything that a rule author needs in order to write a rule should be kept in one place. I'm envisioning a world where a project can write its own rules if it needs to, so it seems useful to keep functions that might be useful in a rule in one place so they're more easily accessible. To me, this includes `fileExists`, `fileConforms`, `directoryConforms`, and `combineRuleExecutionResults`, but also technically `pass` and `fail`, which "end" a rule. The first three we could rightly call "validators" or something, but I'm not sure what name to call the rest. However, we should at least keep `pass` and `fail` with the rest. I've made that change here: [0bd735a](https://github.com/MetaMask/module-lint/pull/39/commits/0bd735ac5492dc08fa694fcc46b8984c29fab281) and here: [c4eb8cd](https://github.com/MetaMask/module-lint/pull/39/commits/c4eb8cd27f3e344c13c11af8bb486aca4959432f)

File: src/rules/helpers.test.ts (hunk 4 above)
Comment (mcmire) [reply src/rules/helpers.test.ts]: I made a change in `utils` which involved writing tests using `withinSandbox` and there was a suggestion there to use `expect.assertions`: https://github.com/MetaMask/utils/pull/148#discussion_r1362700719. I'm half and half on this, so if you don't think this is necessary, that's fine. It sounds like you disagree, so I've dropped this here but let me know if you want it back: [401adc3](https://github.com/MetaMask/module-lint/pull/39/commits/401adc3a411e2765111fbc4ba7dd07f0dd291b48)

File: src/rules/readme-lists-correct-yarn-version.ts (hunk 6 above)
Comment (mcmire) [reply src/rules/readme-lists-correct-yarn-version.ts:16]: You're right! I get a bit paranoid when it comes to regexes, I guess, but this should work just fine. Applied.

File: src/main.test.ts (hunk 3 above)
Comment (mcmire) [reply src/main.test.ts]: Okay, good point. I guess instead of maximal/minimal what I really want is all-passing/all-failing. I've fixed the tests to honor this intent instead and renamed them to reference that better: [65c22a3](https://github.com/MetaMask/module-lint/pull/39/commits/65c22a3092a27d425ad43777bea829bec5e84fdd)

File: src/rules/helpers.test.ts (hunk 4 above)
Comment (Gudahtt) [reply src/rules/helpers.test.ts]: I just didn't understand the reason. I guess it's because the test might end up doing nothing if the helper function we're using here malfunctions? It wasn't totally clear.

File: src/rules/helpers.ts [hunk 8]
```diff
@@ -53,3 +54,165 @@ export function buildRule<Name extends RuleName>({
...
+ * exist in the template exists in the project as well, but also whether all
+ * files in that directory in the template are present in the project and match
+ * content-wise.
+ *
+ * @param directoryPath - The path to a directory in both the template and
+ * project.
+ * @param ruleExecutionArguments - Rule execution arguments.
+ * @returns Either a successful or failed result.
+ */
+export async function directoryConforms(
```
[the remark this answers is not in the document]
Comment (mcmire) [reply src/rules/helpers.ts]: Yeah that's fair! I've made this change here: [404f75f](https://github.com/MetaMask/module-lint/pull/39/commits/404f75f32d49e07bd9c71c0c9cf259d5159a954f)

File: src/rules/readme-lists-correct-yarn-version.ts [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    );
+    if (!match?.[0]) {
+      throw new Error(
```
[the remark this answers is not in the document]
Comment (mcmire) [reply src/rules/readme-lists-correct-yarn-version.ts]: My thought was a rule should only fail if the project does not meet whatever condition is being checked. In this case, we can't even run the rule, because the thing we're using as a point of comparison doesn't exist in the module template. I feel like this is a bug in the rule and not a deficiency in the project itself, so throwing an error seemed more appropriate so that it stood out and hopefully we could be notified of the problem and fix the bug.

That said, you are totally right that throwing an error would prevent other rules from being run, which doesn't seem ideal. I would probably need to fix this in `execute-rules.ts`.

Printing a warning (and returning a passing result) seems like a good solution for now, and then I can follow this up with another PR to properly address this. I've made this change here: [ae3da2c](https://github.com/MetaMask/module-lint/pull/39/commits/ae3da2c2a6fe457ad299c11a6a4ce3c7093a1c66)

File: src/rules/require-valid-package-manifest.ts [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    try {
+      await project.fs.readJsonFileAs(entryPath, PackageManifestSchema);
+      return pass();
+    } catch (error) {
+      if (
+        isErrorWithCode(error) &&
+        isErrorWithMessage(error) &&
+        error.code === 'ERR_INVALID_JSON_FILE'
```
[the remark this answers is not in the document]
Comment (mcmire) [reply src/rules/require-valid-package-manifest.ts:30]: See comment here: https://github.com/MetaMask/module-lint/pull/39#discussion_r1442292663. I've changed this to not throw here: [e1fe878](https://github.com/MetaMask/module-lint/pull/39/commits/e1fe87870bfb5ab651755f12b87ab5cc86e82eab)

[the remark this answers is not in the document]
Comment (Gudahtt) [reply src/rules/require-valid-package-manifest.ts:30]: Sorry I'm not following - why would we want this case to pass?

File: src/rules/helpers.test.ts (hunk 4 above)
Comment (mcmire) [reply src/rules/helpers.test.ts]: That was my interpretation, yes.

File: src/rules/require-valid-package-manifest.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (mcmire) [reply src/rules/require-valid-package-manifest.ts:30]: We don't really want the rule to pass in this case. But we also don't want it to return a failed result, because a failed result means that the project didn't meet a particular condition. At the line highlighted here, we haven't gotten to that point yet, as we're confirming that the module template has a manifest. It _should_, but `readJsonFileAs` can throw, so we need to account for the error. What @kanthesha pointed out above is that if a rule throws, no other rules for the project will get run.

We will have to come back later to fix the `executeRuleNodes` function in `execute-rules.ts` to catch errors as it's iterating through rules, but we have two choices now: we could either continue to throw the error, or not throw it now, "pass" the rule, and then change it back once we know it's safe to throw.

[the remark this answers is not in the document]
Comment (mcmire) [reply src/rules/require-valid-package-manifest.ts:30]: I guess now that I think about it more, it would be more accurate to throw the error now instead of passing.

[the remark this answers is not in the document]
Comment (Gudahtt) [reply src/rules/require-valid-package-manifest.ts:30]: The change made in e1fe87870bfb5ab651755f12b87ab5cc86e82eab pertains to the error case that was already caught and not thrown. I had interpreted @kanthesha 's comment as being about the throwing case (i.e. line 35, as referenced). I'm unsure how the `return fail(...)` could have prevented other rules from running.

File: src/rules/readme-lists-correct-yarn-version.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (mcmire) [reply src/rules/readme-lists-correct-yarn-version.ts:17]: Sorry — reverted in [1b404db](https://github.com/MetaMask/module-lint/pull/39/commits/1b404dbb0a1e0c4ce64ce4f811d91955ccf4d5da) and [d2afd04](https://github.com/MetaMask/module-lint/pull/39/commits/d2afd043fa867da11048ef5d14c1c8358b3db623). For now, I think throwing is more accurate than printing a warning and then passing, which could be confusing without context as to why that's happening. I will still make a ticket to fix `execute-rules.ts`, however.

File: src/rules/require-valid-package-manifest.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (mcmire) [reply src/rules/require-valid-package-manifest.ts:30]: Yes, you're right, and I see what you were talking about now. A hasty fix on my part.

File: src/rules/helpers.ts (hunk 5 above)
Comment (mcmire) [reply src/rules/helpers.ts]: Tests added here: [b683cd0](https://github.com/MetaMask/module-lint/pull/39/commits/b683cd0656ce5966a3e022ecfd9b282ae28cda58)

**Comment by mcmire** [comment]
I believe I've addressed all of the feedback now.

**Review by Gudahtt — DISMISSED** [review]
LGTM!

**Review by Gudahtt — APPROVED** [review]

File: src/rules/all-yarn-modern-files-conform.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (Gudahtt) [reply src/rules/all-yarn-modern-files-conform.ts:16]: Agreed that this is a really nice abstraction! All of the rules turned out quite concise and easy to read.
